### PR TITLE
Implement audit log interceptor

### PR DESCRIPTION
**Pull request by anandjakhaniya**

### CODE REVIEW

File: backend/src/main/java/ai/verta/modeldb/App.java [hunk 1]
```diff
@@ -214,6 +215,7 @@ public static void main(String[] args) {
       serverBuilder.intercept(new ExceptionInterceptor());
       serverBuilder.intercept(new MonitoringInterceptor());
       serverBuilder.intercept(new AuthInterceptor());
+      serverBuilder.intercept(new AuditLogInterceptor());
```
Comment (evgeniilezhnin) [inline backend/src/main/java/ai/verta/modeldb/App.java]: Should also be added in TestsInit

File: backend/src/main/java/ai/verta/modeldb/monitoring/AuditLogInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import io.prometheus.client.Counter;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class AuditLogInterceptor implements ServerInterceptor {
+
+  private static final Logger LOGGER = LogManager.getLogger(AuditLogInterceptor.class);
+  public static final Context.Key<AtomicInteger> CALL_COUNT = Context.key("audit_count");
+  public static final AtomicInteger auditLogCount = new AtomicInteger(0);
```
Comment (evgeniilezhnin) [inline backend/src/main/java/ai/verta/modeldb/monitoring/AuditLogInterceptor.java]: this should have instance for every endpoint call (to be created in interceptCall method)

Comment (conradoverta) [reply backend/src/main/java/ai/verta/modeldb/monitoring/AuditLogInterceptor.java]: Yep

**Review by evgeniilezhnin — APPROVED** [review]

File: backend/common/src/main/java/ai/verta/modeldb/common/monitoring/AuditLogInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  }
+
+  public void increaseAuditCount() {
+    increaseAuditCountStatic();
+  }
+
+  public static void increaseAuditCountStatic() {
+    AtomicInteger atomicInteger = CALL_COUNT.get();
+    if (atomicInteger == null) {
+      LOGGER.warn("Non grpc call to increaseAuditCount");
```
Comment (conradoverta) [inline backend/common/src/main/java/ai/verta/modeldb/common/monitoring/AuditLogInterceptor.java:91]: This is an error, IMO

File: backend/src/main/java/ai/verta/modeldb/audit_log/AuditLogLocalDAORdbImpl.java [hunk 4]
```diff
@@ -31,5 +32,6 @@ public void saveAuditLogs(List<AuditLogLocalEntity> auditLogEntities) {
 
   private void saveAuditLogs(Session session, List<AuditLogLocalEntity> auditLogEntities) {
     auditLogEntities.forEach(session::save);
+    AuditLogInterceptor.increaseAuditCountStatic();
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/audit_log/AuditLogLocalDAORdbImpl.java:38]: This allows multiple entities to be added at once and increase just once. It means that a single API can create multiple entries and still pass the interceptor check. That is the incorrect behavior.

**Review by conradoverta — CHANGES_REQUESTED** [review]

File: backend/common/src/main/java/ai/verta/modeldb/common/monitoring/AuditLogInterceptor.java (hunk 3 above)
Comment (evgeniilezhnin) [reply backend/common/src/main/java/ai/verta/modeldb/common/monitoring/AuditLogInterceptor.java:91]: There are such calls currently on login

Comment (evgeniilezhnin) [reply backend/common/src/main/java/ai/verta/modeldb/common/monitoring/AuditLogInterceptor.java:91]: it doesn't use grpc, but changes user

File: backend/src/main/java/ai/verta/modeldb/audit_log/AuditLogLocalDAORdbImpl.java (hunk 4 above)
Comment (evgeniilezhnin) [reply backend/src/main/java/ai/verta/modeldb/audit_log/AuditLogLocalDAORdbImpl.java:38]: https://github.com/VertaAI/modeldb/pull/1998/

**Pull request merged**